### ISSUE TITLE
[ENH]: restrict x and y does not move out of the range of the graph or floating on the axis when panning and zooming.

### ISSUE THREAD

**Issue by LinqingZ** [matplotlib/matplotlib#26442]
### Problem

I want the graph of heatmap not floating around the on x and y axis, I hope I can zoom in an out only on the graph area, when I zoom out, it becomes so small heatmap with whitespace. I don't want to zoom too much out, set limit to the heatmap range.

```
import matplotlib
import matplotlib.pyplot as plt
import mplcursors
import numpy as np
import torch

def draw_heatmap(heatmap_data):
    figure = plt.figure()
    axes = figure.add_subplot(111)
    heatmap = axes.imshow(heatmap_data, cmap='coolwarm', interpolation='nearest', aspect=1)
    plt.gca().figure.canvas.mpl_connect('scroll_event', lambda event: zoom_heatmap(axes, event))
    plt.show()

def zoom_heatmap(axes, event):
    xlim = axes.get_xlim()
    ylim = axes.get_ylim()
    
    if event.xdata is not None and event.ydata is not None:
        scale_factor = 1.5 if event.button == 'up' else 1/1.5
        new_xlim = (
            (xlim[0] - event.xdata) * scale_factor + event.xdata,
            (xlim[1] - event.xdata) * scale_factor + event.xdata
        )
        new_ylim = (
            (ylim[0] - event.ydata) * scale_factor + event.ydata,
            (ylim[1] - event.ydata) * scale_factor + event.ydata
        )
        
    axes.set_xlim(new_xlim)
    axes.set_ylim(new_ylim)
    axes.figure.canvas.draw()

data = torch.randn(512, 512).numpy()
draw_heatmap(data)
```

Similar issues for the histogram
https://stackoverflow.com/questions/76831010/python-matplotlib-forcing-pan-zoom-to-constrain-the-x-axis-to-a-range-of-x-val

**Comment by stevezhang1999** [comment]
Hi @LinqingZ , can you talk more about what feature do you want? From my understanding, you want to limit the x and y within some range. For example, if we have a 512 by 512 image and you limit both x and y within [-512, 512], then we can only zoom out 0.5x. Also, when we are panning, the x and y will be limited to [-512, 512]. Did I get all you want?

**Comment by tacaswell** [comment]
The logic for updating the view limits based on user input lives at https://github.com/matplotlib/matplotlib/blob/1bea95f1470937ef60d12857d7490008d21e261e/lib/matplotlib/backend_bases.py#L3086-L3125 and https://github.com/matplotlib/matplotlib/blob/1bea95f1470937ef60d12857d7490008d21e261e/lib/matplotlib/axes/_base.py#L4289-L4309 (which is overridden in some Axes subclass).  Both of which will require a bit of chasing to get to the bottom of the implementation. 

On one hand, implementing this is technically straight forward: we need to add some logic to clip the updated limits set.  However, I think this is conceptually/naming-wise complicated as it will require adding a set of "meta-limits" (interactive limits) to the Axes.  We would also have to sort out which (if any) plotting methods automatically set this for the user.

Typing this out I went from "impossible" to "messy" to "mostly requires settling on names + writing documentation".

I think we should consider adding this feature.
